correction-station/correction_station.go

```diff
@@ -0,0 +1,304 @@
+	default:
+		return nil, errStationValidation
+	}
+
+	return deps, nil
+}
+
+// ValidateI2C ensures all parts of the config are valid.
+func (cfg *I2CConfig) ValidateI2C(path string) error {
+	if cfg.I2CBus == "" {
+		return utils.NewConfigValidationFieldRequiredError(path, "i2c_bus")
+	}
+	if cfg.I2cAddr == 0 {
+		return utils.NewConfigValidationFieldRequiredError(path, "i2c_addr")
+	}
+
+	return nil
+}
+
+// ValidateSerial ensures all parts of the config are valid.
+func (cfg *SerialConfig) ValidateSerial(path string) error {
+	if cfg.SerialPath == "" {
+		return utils.NewConfigValidationFieldRequiredError(path, "serial_path")
+	}
+	return nil
+}
+
+type rtkStation struct {
+	resource.Named
+	resource.AlwaysRebuild
+	logger           golog.Logger
+	correctionSource correctionSource
+	protocol         string
+	i2cPath          i2cBusAddr
+	serialWriter     io.Writer
+
+	cancelCtx               context.Context
+	cancelFunc              func()
+	activeBackgroundWorkers sync.WaitGroup
+
+	err movementsensor.LastError
+}
+
+type correctionSource interface {
+	Reader() (io.ReadCloser, error)
+	Start(ready chan<- bool)
+	Close(ctx context.Context) error
+}
+
+type i2cBusAddr struct {
+	bus  board.I2C
+	addr byte
+}
+
+func newRTKStation(
+	ctx context.Context,
+	deps resource.Dependencies,
+	name resource.Name,
+	newConf *StationConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+
+	r := &rtkStation{
+		Named:      name.AsNamed(),
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+		err:        movementsensor.NewLastError(1, 1),
+	}
+
+	r.protocol = newConf.Protocol
+
+	// Init correction source
+	switch r.protocol {
+	case serialStr:
+		err := ConfigureBaseRTKStation(newConf)
+		if err != nil {
+			r.logger.Info("rtk base station could not be configured")
+			return r, err
+		}
+		r.correctionSource, err = newSerialCorrectionSource(newConf, logger)
+		if err != nil {
+			return nil, err
+		}
+		// set a default baud rate if not specficed in config
+		if newConf.SerialBaudRate == 0 {
+			newConf.SerialBaudRate = 38400
+		}
+
+		options := serial.OpenOptions{
+			PortName:        newConf.SerialPath,
+			BaudRate:        uint(newConf.SerialBaudRate),
+			DataBits:        8,
+			StopBits:        1,
+			MinimumReadSize: 4,
+		}
+
+		port, err := serial.Open(options)
+		if err != nil {
+			return nil, err
+		}
+
+		r.logger.Debug("Init serial writer")
+		r.serialWriter = io.Writer(port)
+	case i2cStr:
+		//TODO RSDK-3755 add i2c to this
+	default:
+		// Invalid protocol
+		return nil, fmt.Errorf("%s is not a valid correction source protocol", r.protocol)
+	}
+
+	r.logger.Debug("Starting")
+
+	r.Start(ctx)
+	return r, r.err.Get()
+}
+
+// Start starts reading from the correction source and sends corrections to the radio/bluetooth.
+func (r *rtkStation) Start(ctx context.Context) {
```

Comment:
try to make most of the functions that are driver specific unexported as you go along. Not in this pr but as you edit this module further.